### ISSUE TITLE
[FLUSH CHUNKS]: Unable to find default in Webpack chunks. Please check usage of Babel plugin.

### ISSUE THREAD

**Issue by davidfurlong** [faceyspacey/webpack-flush-chunks#39]
I get this error upon server render. I have no clue what default is supposed to be.

I have resolved the issue by completely rewriting my setup using https://github.com/madeagency/reactivity 

However this error is unclear to me, and perhaps to others. What is meant/ how to debug?

**Comment by faceyspacey** [comment]
The babel-plugin isn't being applied. My best suggestion is to get your setup to look as close to the `universal-demo` as possible, and as you get closer to it, you'll find the culprit.

**Comment by davidfurlong** [comment]
I have a strange variation of this bug where one of my chunks is not being generated, and is thus not found. The strange part of it is that the way its being treated is identical to the others.
Specifically the error: `[FLUSH CHUNKS]: Unable to find containers/Anyone/Home/Home in Webpack chunks. Please check usage of Babel plugin.` is being logged because the file doesnt exist. But in my code every other adjacently `universal-react-component` imported file is being generated:
```
const App = universal(import('./containers/App'), universalOptions);
const Home = universal(
  import('./containers/Anyone/Home/Home'),
  universalOptions
);
const About = universal(
  import('./containers/Anyone/About/About'),
  universalOptions
);
```

**This is very likely an error on my side, so my question is what are the likely sources of this unexpected failure, and are those documented somewhere to make it easier to debug for others in the future?**

**Comment by davidfurlong** [comment]
**RESOLVED** => Dont use `webpack.optimize.AggressiveMergingPlugin()` with this library! (Doh)

**Comment by faceyspacey** [comment]
Excellent!

**Comment by kettanaito** [comment]
@faceyspacey Hi. I'm facing the very same issue, however, I have `babel-plugin-universal-import` in `.babelrc` for my server bundle. I'm not using any aggressive merge plugins.

**server/.babelrc:**
```json
{
  "presets": [
    ["env", {
      "modules": false,
      "targets": {
        "node": "current"
      }
    }],
    "stage-2",
    "react"
  ],
  "plugins": [
    "transform-decorators-legacy",
    "transform-export-extensions",
    ["universal-import", {
      "babelServer": true
    }],
    ["styled-components", {
      "ssr": true
    }]
  ]
}
```

**Client side which server is rendering:**
```jsx
import React from 'react';
import { Switch, Route } from 'react-router-dom';
import universal from 'react-universal-component';

const Frontpage = universal(import('./front-page'));
const About = universal(import('./about'));
const Blog = universal(import('./blog'));
const Page404 = universal(import('./404'));

const RootRoutes = () => (
  <Switch>
    <Route path="/" exact component={ Frontpage } />
    <Route path="/about" component={ About } />
    <Route path="/blog" component={ Blog } />
    <Route path="*" component={ Page404 } />
  </Switch>
);

export default RootRoutes;
```

Getting the same error:
```
[FLUSH CHUNKS]: Unable to find default in Webpack chunks. Please check usage of Babel plugin.
```

Could you please advise me what is going wrong? Thanks.

**Comment by davidfurlong** [comment]
@kettanaito I use the same babel config

```
{
  "presets": [
    ["env", {
      "modules": false,
      "targets": {
        "uglify": true,
        "browsers": ["last 2 versions", "ie >= 10"]
      }
    }],
    "react",
    "stage-0",
    "jest"
  ],
  "compact": true,
  "env": {
    "production": {
      "only": [
        "src"
      ],
      "plugins": [
        "transform-react-remove-prop-types",
        "transform-react-constant-elements",
        "transform-react-inline-elements",
        "transform-flow-strip-types",
      ]
    }
  },
  "plugins": [
    "universal-import",
    "transform-runtime",
    "transform-decorators-legacy",
    "transform-react-display-name",
    "transform-react-stateless-component-name"
  ]
}
```

**Comment by kettanaito** [comment]
@davidfurlong thanks for reply.

The thing is, I have my pages components wrapped in `universal`. This results into error I've mentioned above, and not working SSR. If I remove the wrapper, and pass components directly to the `Route` component - it works beautifully.

All my pages are written like so: `export default class PageName extends React.Component`, all are valid React components. I think the issue is in the `universal()` around them, and that something is most likely missing in order for it to work properly on the server-side.

Is this the `.babelrc` you use to bundle your server or client?

**Comment by davidfurlong** [comment]
that babelrc is for both server and client builds.

**Comment by davidfurlong** [comment]
It might help if you show the contents of your build folder. Could be that you are not using webpack's commonchunks plugin correctly for the client

```
new webpack.optimize.CommonsChunkPlugin({
    names: ['bootstrap'],                   // notice there is no "bootstrap" named entry
    filename: '[name].js',
    minChunks: Infinity
  })
```

**Comment by kettanaito** [comment]
@davidfurlong here's my `client.webpack.config.js`:
```js
/* Webpack */
import webpack from 'webpack';
import BabelMinifyPlugin from 'babel-minify-webpack-plugin';
import OptimizeJsPlugin from 'optimize-js-plugin';
import Stylelint from 'stylelint-webpack-plugin';
import cssnext from 'postcss-cssnext';

/* Utils */
import { merge } from '../webpack.utils';
import { providers, processors, resolvers } from '../webpack.parts';
import { forEnv, absolutePath } from '../../utils';
import { CONFIG, CLIENT, SHARED, VENDOR, WEBPACK } from '../../../config';

/* Shorthand: Environment */
const DEVELOPMENT = (process.env.NODE_ENV === 'development');
const PRODUCTION = (process.env.NODE_ENV === 'production');

export default merge([
  {
    entry: {
      client: ['babel-polyfill', CLIENT.SRC_ENTRY]
    },
    output: {
      /* Output directory to compile to the disk */
      path: absolutePath(CLIENT.BUILD_DIR),

      /* Include modules information in the bundle */
      pathinfo: DEVELOPMENT,

      /**
       * Reference path to access the bundle.
       * webpack-dev-server makes bundle available under this path from the memory.
       */
      publicPath: CLIENT.BUILD_DIRNAME,

      /* File name of compiled bundle */
      filename: DEVELOPMENT ? 'index.js' : 'index.[chunkhash].js',

      /* File name format of each compiled chunk */
      chunkFilename: `${WEBPACK.CHUNKS_DIRNAME}/${DEVELOPMENT ? '[name]' : '[name].[chunkhash]'}.js`
    },
    plugins: forEnv({
      any: [
        new webpack.DefinePlugin({
          'process.env.NODE_ENV': JSON.stringify(process.env.NODE_ENV),
          'process.env.BABEL_ENV': JSON.stringify(process.env.NODE_ENV),
          'window.fetch': null // Disable native fetch to use Bluebird instead
        }),
        new webpack.DllReferencePlugin({
          context: process.cwd(),
          manifest: absolutePath(VENDOR.MANIFEST_FILEPATH)
        }),
        new webpack.optimize.CommonsChunkPlugin({
          names: ['webpack-bootstrap'],
          filename: '[name].js',
          minChunks: Infinity
        }),
        new webpack.LoaderOptionsPlugin({
          options: {
            postcss: [cssnext()]
          }
        }),
        new Stylelint({ syntax: 'scss' })
      ],

      development: [
        /* Use verbose module names */
        new webpack.NamedModulesPlugin(),

        /* Enable hot module replacement */
        new webpack.HotModuleReplacementPlugin(),

        /* Use loaders in debug mode */
        new webpack.LoaderOptionsPlugin({
          debug: true
        })
      ],

      production: [
        /* Ensure scope hoisting */
        new webpack.optimize.ModuleConcatenationPlugin(),

        /* Optimize the code for browser usage */
        new OptimizeJsPlugin({
          sourceMap: false
        }),

        /* Minify and tree shake */
        new BabelMinifyPlugin({
          removeConsole: true,
          removeDebugger: true,
          mangle: {
            topLevel: true
          }
        }),

        /* Prevent writing files to disk on error */
        new webpack.NoEmitOnErrorsPlugin()
      ]
    }),
    cache: true,
    profile: true,
    devtool: DEVELOPMENT ? 'eval' : 'cheap-module-inline-source-map',
    node: {
      fs: 'empty',
      net: 'empty',
      tls: 'empty'
    },
    performance: {
      /* Increase maximum entry/asset file size due to Gzip compression on the server */
      maxEntrypointSize: 1000000,
      maxAssetSize: 1000000,
      hints: PRODUCTION && 'warning'
    }
  },

  /* Provide global modules references */
  providers(),

  /* JavaScript */
  processors.js({
    include: [
      absolutePath(CONFIG.DIR),
      absolutePath(CLIENT.SRC_DIR),
      absolutePath(SHARED.SRC_DIR)
    ],
    hot: DEVELOPMENT
  }),

  /* GraphQL */
  processors.graphql({
    include: [
      absolutePath(CLIENT.SRC_DIR)
    ]
  }),

  /* CSS-in-JS */
  processors.css({
    include: [
      absolutePath(CLIENT.SRC_DIR)
    ]
  }),

  /* Provide custom resolvers */
  resolvers()
]);
```

I indeed bundle the bootstrap part into `webpack-bootstrap` bundle and specify this in `flushChunks`:
```js
const { Js: JsChunks } = flushChunks(clientStats, {
      chunkNames,
      before: ['webpack-bootstrap', 'vendor', 'client']
    });
```

I don't bundle `vendor`, though. That was another question, but shouldn't be raising this issue. Otherwise the error message is completely not pointing where the error happens.

**Comment by davidfurlong** [comment]
> filename: DEVELOPMENT ? 'index.js' : 'index.[chunkhash].js',

So I'm not expert (but I have grappled with similar webpack issues)

Try disabling some of your Plugins, they may be renaming / concatenating - particularly moduleContatenationPlugin. Also maybe the DLL plugin just to see if thats the issue

Also transform-export-extensions babel plugin could be renaming your exports 

Also if you log out the contents of your build directory post build it might help

**Comment by kettanaito** [comment]
```
filename: DEVELOPMENT ? 'index.js' : 'index.[chunkhash].js',
```
That just means that production bundle name will include the chunkhash. There is no issue in loading the client entry, even with the chunkhash. I don't think that's the issue, unless `babel-plugin-universal-import` utilizes that in some weird way. Moreover, the flushed JS chunks include the proper bundle name, with the hash. The issue I'm experiencing is related to the chunks generated by the `universal(import())` split points.

`moduleContatenationPlugin` might have been a good catch, but it's applied only in development. Unfortunately, I'm experiencing the issue in production build.

`transform-export-extensions` hm. My codebase heavily relies on `export foo from './foo` statement without explicit imports. I would try to disable this plugin as the very last step, but thanks, this is valuable.

## Here's the build folder after build process:
![Image](https://image.prntscr.com/image/f3EsVhsVQxaYw0WQueQBYQ.png)

**Comment by kettanaito** [comment]
Removing `moduleContatenationPlugin` *changed* things in some way.

Now I see no error in the console, but the layout I get from SSR contains `<div>Loading...</div>` instead of the loaded, rendered component which I wrapped in `universal()`. That causes client-side `ReactDOM.hydrate()` to fail, since it's not expecting that node.

Any ideas on that?

**Comment by davidfurlong** [comment]
I've hit the limit of my knowledge. All I can advise is to remove plugins
until our configs are the same.

On Sun, Feb 4, 2018 at 6:09 PM, Artem Zakharchenko <notifications@github.com

**Comment by kettanaito** [comment]
@davidfurlong thanks for the help anyway.

Breaking it down to the identical configs is a valid approach, but I have a relatively big app where I migrate from react-router 3 to 4, re-writing SSR by the road. I cannot change how the app bundles too much, that would mean spending weeks on re-writing the whole app's architecture just to see if that helps.

Sorry, if some technology is implemented according to the official evident docs, and still not working, redoing the present and working architecture as a part of troubleshooting makes me think something is not very clear with the technology itself.

**Comment by kettanaito** [comment]
## Solution
At least, what fixed this issue for me was providing a proper `whitelist` option to `nodeExternals` in `server.webpack.config.js`:

```js
{
  externals: [
    nodeExternals({
      whitelist: [
        /babel-plugin-universal-import|react-universal-component/
      ]
    })
  ]
}
```
> **Note:** Some examples in the discussions in `react-universal-component` suggest passing a function to `whitelist` array, however, its types say it only accepts strings or RegExp.

The point about `nodeExternals()` should really be highlighted big bold and red, as it seems to be the reason for issues for many people over here. Thanks for help.

**Comment by faceyspacey** [comment]
@kettanaito glad you were able to get it to work. The nodeExternals package has proved problematic for many, which is why the recommended approach has been to do it manually:

https://github.com/faceyspacey/webpack-flush-chunks#externals

It seems you have pinpointed the proper way to use nodeExternals.

**Comment by martinchristov** [comment]
I'm experiencing this variation of the issue:
`[FLUSH CHUNKS]: Unable to find views/home/checkout/thankyou in Webpack chunks. Please check usage of Babel plugin.`

I discovered that the problem i do this:

`const UniView = universal(({view}) => import(`./views/${view}`));`
`<UniView view="home/checkout/thankyou" />"`

The issue is resolved if the above is changed to:
`const UniView = universal(({view}) => import(`./views/home/checkout/${view}`));`
`<UniView view="thankyou" />"`

Not sure why is that a problem.

**Comment by DannyDane** [comment]
I'm having the same issue.


I found out that for some reason 2 request are made to the server. For the first request chunknames is an array with the correct chunks, but on the second request the chunknames array is empty.
That's why it's complaining for missing chunks.

Anyone experiencing the same problem might want to try console logging the chunknames variable on the server side.

No idea how to fix the 2 request to the server issue though

**Comment by faceyspacey** [comment]
@DannyDane you must be using Apollo server rendering or doing something else async on the server. If that is in fact the issue, `clearChunks` will solve it:

https://github.com/faceyspacey/react-universal-component#example-with-react-apollo-and-aphrodite

**Comment by DannyDane** [comment]
**TLDR:** It had nothing to do with webpack-flush-chunks

Yeah, the double rendering had nothing to do with webpack-flush-chunks, my fault.
But even after I fixed the double rendering the error was still there.

Interestingly enough, it was because I was importing the component from a folder (ie: import('folder/MyComponent') and in that same folder I had an index.js file that exported that same component (export { default as MyComponent } from './MyComponent)

And in the same file where I was dynamically importing that specific component I was also statically importing other components from that same folder using destructering (import { OtherComponent } from folder).

And I forgot that importing files using that method imports all the files, not only the files you put between the brackets.

**Comment by aneogiarcadix** [comment]
I got the same error in a Typescript project of mine. So I cloned the universal-demo project and changed things to typescript( except server/index.js ). The project builds successfully, and then I get the same error. 
![image](https://user-images.githubusercontent.com/31309583/40896578-5cfb7472-67d3-11e8-951e-1389cea4c148.png)
What am I missing here? The only changes I have made to the webpack configs is adding at-loader to the rules and including ts and tsx to resolve extensions. 

Current project structure:
![image](https://user-images.githubusercontent.com/31309583/40896664-ea494426-67d3-11e8-99e2-e0ccff89de52.png)
tsconfig.json
{
    "compileOnSave": true,
    "exclude": ["node_modules"],
    "compilerOptions": {
      "noStrictGenericChecks": true,
      "allowJs": true,
      "declaration": false,
      "noResolve": false,
      "jsx": "react",
      "module": "commonjs",
      "target": "es2015", 
      "experimentalDecorators": true,
      "sourceMap": true,
      "noImplicitAny": false,
      "noImplicitReturns": true,
      "outDir": "./dist/",
      "skipLibCheck": true,
      "allowSyntheticDefaultImports": true,
      "moduleResolution": "node",
      //"types": [ "node" ],
      "typeRoots": []
    }
}

**Comment by zackljackson** [comment]
https://github.com/faceyspacey/webpack-flush-chunks/pull/59

Now available via `@next` tag on NPM

**Comment by davidfurlong** [comment]
@zackljackson I'm super impressed and delighted by all your work on this ecosystem
